tests/python/pants_test/option/test_options_fingerprinter.py

```diff
@@ -5,17 +5,36 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
 import os
+import unittest
 from builtins import range, zip
 
 from pants.base.payload import Payload
 from pants.base.payload_field import PrimitiveField
 from pants.option.custom_types import (UnsetBool, dict_option, dir_option, file_option, list_option,
                                        target_option)
-from pants.option.options_fingerprinter import OptionsFingerprinter
+from pants.option.options_fingerprinter import OptionsFingerprinter, stable_json_sha1
 from pants.util.contextutil import temporary_dir
 from pants_test.test_base import TestBase
 
 
+class JsonEncodingTest(unittest.TestCase):
+  def test_known_checksums(self):
+    """Check a laundry list of supported inputs to stable_json_sha1().
+
+    This checks both that the method can successfully handle the type of input object, but also that
+    the hash of specific objects remains stable.
+    """
+    self.assertEqual(stable_json_sha1({}), 'bf21a9e8fbc5a3846fb05b4fa0859e0917b2202f')
+    self.assertEqual(stable_json_sha1(()), '97d170e1550eee4afc0af065b78cda302a97674c')
+    self.assertEqual(stable_json_sha1([]), '97d170e1550eee4afc0af065b78cda302a97674c')
+    self.assertEqual(stable_json_sha1(set([])), '97d170e1550eee4afc0af065b78cda302a97674c')
+    self.assertEqual(stable_json_sha1([{}]), '4e9950a1f2305f56d358cad23f28203fb3aacbef')
+    self.assertEqual(stable_json_sha1([('a', 3)]), 'd6abed2e53c1595fb3075ecbe020365a47af1f6f')
+    self.assertEqual(stable_json_sha1({'a': 3}), '9e0e6d8a99c72daf40337183358cbef91bba7311')
+    self.assertEqual(stable_json_sha1([{'a': 3}]), '8f4e36849a0b8fbe9c4a822c80fbee047c65458a')
+    self.assertEqual(stable_json_sha1(set([1])), 'f629ae44b7b3dcfed444d363e626edf411ec69a8')
```

Comment:
Wonder if it's an option to assert on bytestring instead. Imagine someone breaks one of the tests in the future, how should they start debugging?

Reply:
I'm not sure -- it would have to be a change to either `stable_json_sha1`, `stable_json_hash`, or the `Encoder` class that changes the result of `stable_json_sha1`. What do you mean by assert on bytestring?

Reply:
`stable_json_hash ` is essentially hashing `json_str` which I am guessing is a byte string (e.g. `b'123'` instead of a `str` like `'123'`).

https://github.com/pantsbuild/pants/blob/3be90c2f31a419ef96c65434d17c99b4a56043ef/src/python/pants/base/hash_utils.py#L54-L55

so the assert would look something like
```
self.assertEqual(b'"{"a":3}"', get_json_str({'a': 3}))
```

Reply:
That makes perfect sense. I can totally implement that. I don’t know if separating the hashing from the json-ification will be obvious or take like 5 minutes but I’ll implement this. Thanks!

Reply:
I broke out #6421 because I wasn't exactly sure how to implement this in this PR (splitting out the json encoding from the hashing), but for the reasons you describe I think this should be done.